src/nix-build/nix-build.cc

```diff
@@ -423,13 +436,18 @@ static void _main(int argc, char * * argv)
            lose the current $PATH directories. */
         auto rcfile = (Path) tmpDir + "/rc";
         writeFile(rcfile, fmt(
-                (keepTmp ? "" : "rm -rf '%1%'; "s) +
+                R"(_nix_shell_clean_tmpdir() { rm -rf %1%; }; )"s +
+                (keepTmp ?
+                    "trap _nix_shell_clean_tmpdir EXIT; "
+                    "exitHooks+=(_nix_shell_clean_tmpdir); "
+                    "failureHooks+=(_nix_shell_clean_tmpdir); ":
```

Comment:
[stdenv setup] overrides the trap with `exitHandler` function which calls `exitHook` or `failureHook`. I've put the trap at line 441 as a fallback.

[stdenv setup]: https://github.com/NixOS/nixpkgs/blob/39700a65ef75e473f46270e07238cabff0462631/pkgs/stdenv/generic/setup.sh#L113-L149